crates/rpc/ipc/src/server/mod.rs

```diff
@@ -21,31 +24,50 @@ use tokio::{
     io::{AsyncRead, AsyncWrite},
     sync::{oneshot, watch, OwnedSemaphorePermit},
 };
-use tower::{layer::util::Identity, Service};
+use tower::{layer::util::Identity, Layer, Service};
 use tracing::{debug, trace, warn};
 
 // re-export so can be used during builder setup
-use crate::server::connection::IpcConnDriver;
+use crate::server::{
+    connection::IpcConnDriver,
+    rpc_service::{RpcService, RpcServiceCfg},
+};
 pub use parity_tokio_ipc::Endpoint;
 use tokio::sync::mpsc;
 use tokio_stream::wrappers::ReceiverStream;
+use tower::layer::{util::Stack, LayerFn};
 
 mod connection;
 mod future;
 mod ipc;
+mod rpc_service;
 
 /// Ipc Server implementation
 
 // This is an adapted `jsonrpsee` Server, but for `Ipc` connections.
-pub struct IpcServer<B = Identity> {
+pub struct IpcServer<HttpMiddleware = Identity, RpcMiddleware = Identity> {
     /// The endpoint we listen for incoming transactions
     endpoint: Endpoint,
     id_provider: Arc<dyn IdProvider>,
     cfg: Settings,
-    service_builder: tower::ServiceBuilder<B>,
+    rpc_middleware: RpcServiceBuilder<RpcMiddleware>,
+    http_middleware: tower::ServiceBuilder<HttpMiddleware>,
 }
 
-impl IpcServer<Identity> {
+impl<HttpMiddleware, RpcMiddleware> IpcServer<HttpMiddleware, RpcMiddleware>
+where
+    RpcMiddleware: Layer<RpcService> + Clone + Send + 'static,
+    for<'a> <RpcMiddleware as Layer<RpcService>>::Service: RpcServiceT<'a>,
+    HttpMiddleware: Layer<TowerServiceNoHttp<RpcMiddleware>> + Send + 'static,
+    <HttpMiddleware as Layer<TowerServiceNoHttp<RpcMiddleware>>>::Service: Send
+        + Service<
+            String,
+            Response = Option<String>,
+            Error = Box<dyn std::error::Error + Send + Sync + 'static>,
+        >,
+    <<HttpMiddleware as Layer<TowerServiceNoHttp<RpcMiddleware>>>::Service as Service<String>>::Future:
+    Send + Unpin,
+{
```

Comment:
we can relax this a bit and on require these bounds for functions that need it and don't require anything else for the impl block with functions that don't need it

Reply:
Would like to try it after clean up codes.

Reply:
Is https://github.com/paradigmxyz/reth/pull/7790/commits/110e4c7f08f6430d3d96d9263ef0a99d7d6f676e  what you want?  

`IpcServer::start_inner` require these bounds, and `IpcServer::start` call ``IpcServer::start_inner` so both functions in the impl block need these bounds. So I move `IpcServer::endpoint` out.